src/ucp/wireup/wireup.c

```diff
@@ -419,14 +419,125 @@ ucp_wireup_connect_local(ucp_ep_h ep,
+                                      UCS_CALLBACKQ_FLAG_ONESHOT, &prog_id);
+    ucp_worker_signal_internal(ucp_ep->worker);
+}
+
+/*
+ * We switch the endpoints in this function (instead in wireup code) since
+ * this is guaranteed to run from the main thread.
+ */
+unsigned ucp_wireup_eps_progress(void *arg)
+{
+    ucp_ep_h ucp_ep = arg;
+    ucp_wireup_ep_t *wireup_ep;
+    ucs_queue_head_t tmp_pending_queue;
+    uct_pending_req_t *uct_req;
+    ucp_lane_index_t lane;
+
+    UCS_ASYNC_BLOCK(&ucp_ep->worker->async);
+
+    /* If we still have pending wireup messages, send them out first */
+    if (ucp_ep_has_wireup_pending(ucp_ep)) {
+        ucp_wireup_eps_progress_sched(ucp_ep);
+        goto out_unblock;
+    }
+
+    /* If an error happened on the endpoint (but perhaps the deferred error handler,
+     * ucp_worker_iface_err_handle_progress(), was not called yet, avoid changing
+     * ep state, and let the error handler take care of cleanup.
+     */
+    if (ucp_ep->flags & UCP_EP_FLAG_FAILED) {
+        ucs_trace("ep %p: not switching wireup eps to ready state because of "
+                  "error", ucp_ep);
+        goto out_unblock;
+    }
+
+    /* Move wireup pending queue to temporary queue and remove references to
+     * the wireup progress function
+     */
+    ucs_queue_head_init(&tmp_pending_queue);
+    for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep); ++lane) {
+        wireup_ep = ucp_wireup_ep(ucp_ep_get_lane(ucp_ep, lane));
+        if (wireup_ep == NULL) {
+            continue;
+        }
+
+        ucs_trace("ep %p: switching wireup_ep %p to ready state", ucp_ep,
+                  wireup_ep);
+
+        ucs_queue_for_each_extract(uct_req, &wireup_ep->pending_q, priv, 1) {
+            ucs_queue_push(&tmp_pending_queue, ucp_wireup_ep_req_priv(uct_req));
+            ucp_worker_flush_ops_count_dec(ucp_ep->worker);
+        }
+
+        /* Switch to real transport and destroy proxy endpoint (aux_ep as well) */
+        ucp_proxy_ep_replace(&wireup_ep->super);
+    }
+
+    /* Replay pending requests */
+    ucp_wireup_replay_pending_requests(ucp_ep, &tmp_pending_queue);
+    UCS_ASYNC_UNBLOCK(&ucp_ep->worker->async);
+    return 1;
+
+out_unblock:
+    UCS_ASYNC_UNBLOCK(&ucp_ep->worker->async);
+    return 0;
+}
+
+void ucp_wireup_remote_connect_lanes(ucp_ep_h ucp_ep, int ready)
+{
+    ucp_lane_index_t lane;
+    ucp_wireup_ep_t *wireup_ep;
+
+    for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep); ++lane) {
+        wireup_ep = ucp_wireup_ep(ucp_ep_get_lane(ucp_ep, lane));
+        if (wireup_ep == NULL) {
+            continue;
         }
+
+        ucs_assert(wireup_ep->super.uct_ep != NULL);
+        ucs_assert(wireup_ep->super.ucp_ep == ucp_ep);
+        ucs_assert(wireup_ep->flags & UCP_WIREUP_EP_FLAG_LOCAL_CONNECTED);
+
+        wireup_ep->flags |= UCP_WIREUP_EP_FLAG_REMOTE_CONNECTED;
+        if (ready) {
+            wireup_ep->flags |= UCP_WIREUP_EP_FLAG_READY;
+        }
```

Comment:
maybe rename the func to ucp_wireup_update_flags(ucp_ep, flags)
that will add "flags" to all wireup_ep->flags